File: modal/sandboxsnapshot.py
# Copyright Modal Labs 2024

from typing import Optional

from modal_proto import api_pb2

from ._resolver import Resolver
from ._utils.async_utils import synchronize_api
from ._utils.grpc_utils import retry_transient_errors
from .client import _Client
from .object import _Object


class _SandboxSnapshot(_Object, type_prefix="sn"):
    """A `SandboxSnapshot` object lets you interact with a stored Sandbox snapshot that was created by calling
    .snapshot() on a Sandbox instance. This includes both the filesystem and memory state of the original Sandbox at the
    time the snapshot was taken.
    """

    @staticmethod
    async def from_id(sandbox_snapshot_id: str, client: Optional[_Client] = None):
        if client is None:
            client = await _Client.from_env()

        async def _load(self: _SandboxSnapshot, resolver: Resolver, existing_object_id: Optional[str]):
            resp = await retry_transient_errors(
                client.stub.SandboxSnapshotFromId,
                api_pb2.SandboxSnapshotFromIdRequest(sandbox_snapshot_id=sandbox_snapshot_id),
            )
            self._hydrate(resp.sandbox_snapshot_id, resolver.client, resp.metadata)

        rep = "SandboxSnapshot()"
        obj = _SandboxSnapshot._from_loader(_load, rep)

        return obj


SandboxSnapshot = synchronize_api(_SandboxSnapshot)

File: js/src/sandbox_snapshot.ts
import { type ModalClient } from "./client";

/**
 * > Sandbox memory snapshots are in **early preview**.
 *
 * A `SandboxSnapshot` object lets you interact with a stored Sandbox snapshot that was created by calling
 * {@link Sandbox#experimentalSnapshot} on a Sandbox instance. This includes both the filesystem and memory state of
 * the original Sandbox at the time the snapshot was taken.
 */
export class SandboxSnapshot {
  readonly snapshotId: string;
  readonly #client: ModalClient;
  #isV2: boolean | undefined;

  /** @ignore */
  constructor(
    client: ModalClient,
    snapshotId: string,
    params: { isV2?: boolean } = {},
  ) {
    this.#client = client;
    this.snapshotId = snapshotId;
    this.#isV2 = params.isV2;
  }

  /**
   * @internal
   * Whether the snapshot came from a V2 sandbox. `undefined` until hydrated.
   */
  get _isV2(): boolean | undefined {
    return this.#isV2;
  }

  /** @internal */
  async _hydrate(): Promise<void> {
    if (this.#isV2 !== undefined) {
      return;
    }
    // hydration doesn't actually do much apart from validating the existance of the id
    // which is implicitly done by trying to start a sandbox from the snapshot as well
    const resp = await this.#client.cpClient.sandboxSnapshotGet({
      snapshotId: this.snapshotId,
    });
    this.#isV2 = resp.handleMetadata?.isV2 ?? false;
  }
}

/**
 * Service for managing {@link SandboxSnapshot}s.
 *
 * Normally only ever accessed via the client as:
 * ```typescript
 * const modal = new ModalClient();
 * const snapshot = await modal.sandboxSnapshots.fromId(snapshotId);
 * ```
 */
export class SandboxSnapshotService {
  readonly #client: ModalClient;
  constructor(client: ModalClient) {
    this.#client = client;
  }

  /**
   * Construct a {@link SandboxSnapshot} for an existing snapshot ID.
   *
   * @param snapshotId - Snapshot ID returned when the snapshot was created.
   */
  async fromId(snapshotId: string): Promise<SandboxSnapshot> {
    return new SandboxSnapshot(this.#client, snapshotId);
  }
}
